contracts/deployer/AccessControlledDeployer.sol

```diff
@@ -0,0 +1,165 @@
+        if (address(deployer) == address(0)) {
+            revert ZeroAddress();
+        }
+        return deployer.deploy{value: msg.value}(bytecode, salt);
+    }
+
+    /**
+     * @notice Deploys a contract using a deployment method defined by `deployer` and initializes it
+     * @param deployer The create2 or create3 deployer contract that will deploy the contract
+     * @param bytecode The bytecode of the contract to be deployed
+     * @param salt A salt to influence the contract address
+     * @param init Init data used to initialize the deployed contract
+     * @dev Only address with DEPLOYER_ROLE can call this function
+     * @dev This function requires that the current owner of `deployer` is this contract
+     * @dev The function can only be called if the contract is not in a paused state
+     * @dev The function emits `Deployed` event after the contract is deployed
+     * @return The address of the deployed contract
+     */
+    function deployAndInit(
+        IDeployer deployer,
+        bytes memory bytecode,
+        bytes32 salt,
+        bytes calldata init
+    ) external payable whenNotPaused onlyRole(DEPLOYER_ROLE) returns (address) {
+        if (address(deployer) == address(0)) {
+            revert ZeroAddress();
+        }
+        return deployer.deployAndInit{value: msg.value}(bytecode, salt, init);
+    }
+
+    /**
+     * @notice Grants a list of addresses the DEPLOYER_ROLE
+     * @param deployers list of addresses to grant the DEPLOYER_ROLE
+     * @dev Only address with DEFAULT_ADMIN_ROLE can call this function
+     * @dev The function emits `RoleGranted` event for each address granted the DEPLOYER_ROLE.
+     *      This is not emitted if an address is already a deployer
+     */
+    function grantDeployerRole(address[] memory deployers) public {
+        if (deployers.length == 0) {
+            revert EmptyDeployerList();
+        }
+        for (uint256 i = 0; i < deployers.length; i++) {
+            if (deployers[i] == address(0)) {
+                revert ZeroAddress();
+            }
+            grantRole(DEPLOYER_ROLE, deployers[i]);
+        }
+    }
+
+    /**
+     * @notice Revokes the DEPLOYER_ROLE from a list of addresses
+     * @param deployers list of addresses to revoke the DEPLOYER_ROLE from
+     * @dev Only address with DEFAULT_ADMIN_ROLE can call this function
+     * @dev The function emits `RoleRevoked` event for each address for which the DEPLOYER_ROLE was revoked
+     *      This is not emitted if an address was not a deployer
+     */
+    function revokeDeployerRole(address[] memory deployers) public {
+        if (deployers.length == 0) {
+            revert EmptyDeployerList();
+        }
+        for (uint256 i = 0; i < deployers.length; i++) {
+            if (deployers[i] == address(0)) {
+                revert ZeroAddress();
+            }
+            revokeRole(DEPLOYER_ROLE, deployers[i]);
+        }
```

Comment:
Are there any problems with not ensuring `deployers[i]` actually has the role first, before revoking? I don't think so

Reply:
No, I don't believe so. OZ `AccessControl` performs this check internally in `_revokeRole()` [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v4.9/contracts/access/AccessControl.sol#L242) and ignores the request if the user doesn't have the role